source/core/slang-string.h

```diff
@@ -313,6 +342,8 @@ namespace Slang
 		{
 		}
 
+        SLANG_FORCE_INLINE StringRepresentation* getStringRepresentation() const { return buffer; }
```

Comment:
Note that in principle `StringRepresentation` is supposed to be an implementation detail, and not part of the public API of `String`. In the long term, I'd like to make it so that `String` actually stores a direct `char*` that points *into* the `StringRepresentation` backing storage, so that in a debugger view it visualizes as just the text data (in the 99% case where the content is nul-terminated).

I understand the need to expose the implementation details for now, but I'd actually like to consider cleaning this up at some point in the future (or moving all the access to raw internals of the `String` representation into a `friend` utility class that lives in another header, so people don't see these operations by default).

Reply:
I don't see this as a particular problem in terms of a future update to String, because within the code to hold the reference it can just hold whatever the new representation is. It was a happy (by perhaps somewhat inefficient) coincidence it was RefObject derived. 

If the 'new representation' isn't a pointer, then that will require holding some other rep that copies the data such that it can be ref'd and held as a pointer (or something like that) inside of the IRInst derived class.

Reply:
Yeah, the new `String` implementation I introduced was almost entirely motivated by two goals:

1. I wanted `sizeof(String) == sizeof(void*)`, so that `String`s add minimal storage overhead to things that use them. I also wanted (but didn't implement) to have a `String` be compatible (via aliasing) with a `char*` for cases where you want to expose a raw C-compatible view of things.

2. I wanted `String` to support copy-on-write optimization, so that appending to a `String` could be an efficient operation, thus eliminating most of the need for the separate `StringBuilder` type.

These two goals meant that having a `String` just be a pointer to a reference-counted payload was the best option, and deriving from the existing `RefObject` made sense there. I'd actually like to push `List`/`Array` in a similar direction, because efficient copy-on-write is a good default for a lot of non-performance-critical types, and reducing the actual size of objects that include strings/arrays/etc. in their memory layout seems like a general goodness when many of those fields will end up with empty/default values.

I agree these aren't the most efficient implementation choices (e.g., a good `std:;string` will do all kinds of tricks to avoid heap allocation for small strings), but the main benefit is reducing rampant heap allocation/copying when passing around these types between functions. We can optimize further down the line, but I think copy-on-write is a good default starting point.